src/allmydata/test/test_client.py

```diff
@@ -516,32 +531,42 @@ def _permute(self, sb, key):
 
     def test_permute(self):
         sb = StorageFarmBroker(True, None, EMPTY_CLIENT_CONFIG)
-        for k in ["%d" % i for i in range(5)]:
+        ks = [b"%d" % i for i in range(5)]
+        for k in ks:
             ann = {"anonymous-storage-FURL": SOME_FURL,
                    "permutation-seed-base32": base32.b2a(k) }
             sb.test_add_rref(k, "rref", ann)
 
-        self.failUnlessReallyEqual(self._permute(sb, "one"), ['3','1','0','4','2'])
-        self.failUnlessReallyEqual(self._permute(sb, "two"), ['0','4','2','1','3'])
+        one = self._permute(sb, b"one")
+        two = self._permute(sb, b"two")
+        self.assertEqual(sorted(one), ks)
+        self.assertEqual(sorted(two), ks)
+        self.assertNotEqual(one, two)
         sb.servers.clear()
-        self.failUnlessReallyEqual(self._permute(sb, "one"), [])
+        self.failUnlessReallyEqual(self._permute(sb, b"one"), [])
```

Comment:
I'm not sure this was a great test before but I think it had at least one important property that the new version doesn't have.

The exact order of the list it returns is a core part of the functionality of `StorageFarmBroker.get_servers_for_psi`.

This is because the order of these results determines which servers a client will choose to place shares on and which servers it will consult (and in what order) when trying to retrieve those shares.  If the order ever changes, all already-placed shares become (at best) harder to find or (at worst) impossible to find.

So I'd like to see `['3','1','0','4','2']` and `['0','4','2','1','3']` preserved in the test suite somehow.

Also maybe some of this information belongs in a docstring for this method since clearly the current form of the test doesn't convey any of this informatoin.

Reply:
Urrrgh. I think this may have depended on Python's built-in hash algorithm, which has changed in Python 3, but optimistically I'm misremembering.

Reply:
Ah, no, `sha1`, this won't be too bad.